### DISCUSSION TITLE
Optional deps with at least one required

### DISCUSSION THREAD

**Discussion by k0nserv** [Homebrew/discussions#5795] [Writing Formulae/Casks]
### Output of `brew config`

```shell
N/A
```

### Output of `brew doctor`

```shell
N/A
```

### Description of issue

Hello,

I was looking at the [formula](https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/p/pgvector.rb) for [pgvector](https://github.com/pgvector/pgvector) and wanted to suggest an improvement.

Being an extension to Postgres it needs Postgres to be installed, this is currently achieved by directly depending on `postgresql@14` and `postgresql@17`. I wanted to suggest an improvement to this formula that changes this to optional dependencies e.g.

```ruby
  depends_on "postgresql@13" => :optional
  depends_on "postgresql@14" => :optional
  depends_on "postgresql@15" => :optional
  depends_on "postgresql@16" => :optional
  depends_on "postgresql@17" => :optional
```

However, a user should specify at least one of these optional dependencies.

Thus my questions are:

1. What's the best way to express this "many optional, but at least one required" semantics?
2. Can you bail earlier than in `install` if this condition isn't met?
3. Would it be possible to upgrade `depends_on "postgresql@17" => :optional` to `depends_on "postgresql@17" => [:build, :test]` if the user does not specify any version of Postgres explicitly?

@ankane I'm happy to contribute this if you like the direction

**Comment by SMillerDev** [comment]
Please see https://docs.brew.sh/Formula-Cookbook#specifying-other-formulae-as-dependencies

`:optional` is not allowed in Homebrew/core

**Reply by k0nserv** [reply]
Oh, right. Is there some other way to achieve this, I guess it would still involve forcing the end user to specify some command line switch, which might be against the rules for Homebrew/core for the same reason?

**Reply by SMillerDev** [reply]
Any sort of user defined configuration is disallowed in core because it means that configuration won't be tested in CI

**Reply by k0nserv** [reply]
> Any sort of user defined configuration is disallowed in core because it means that configuration won't be tested in CI

Is it okay to have user configuration, but with a sensible default or is it a total ban?

**Reply by SMillerDev** [reply]
No, because that introduces the exact same problem of untested configurations.

**Reply by k0nserv** [reply]
Alright, thanks for explaining!

It sounds like there's no way to solve the usability problems with pgvector while it resides in `homebrew/core` then?

In my mind those problems are:

1. Redundantly installing postgresql versions the user might not be using(14 and 17 at the moment)
2. Not supporting installation with other Postgres versions(installed via Homebrew), 13, 15, and 16 at the moment

**Reply by ankane** [reply]
Hi @k0nserv, it doesn't install unnecessary Postgres versions (the dependencies are only for build and test). It uses the same pattern as other Postgres extensions, so I don't think any changes are needed (supporting 13, 15, and 16 would break existing Postgres installations - see homebrew-core for previous discussions / attempts at this).

**Reply by SMillerDev** [reply]
> Redundantly installing postgresql versions the user might not be using(14 and 17 at the moment)

This is only the case if the user runs an unsupported brew configuration

> Not supporting installation with other Postgres versions(installed via Homebrew), 13, 15, and 16 at the moment

This seems to be easily solved by adding dependencies on those

**Reply by k0nserv** [reply]
> Hi @k0nserv, it doesn't install unnecessary Postgres versions (the dependencies are only for build and test). It uses the same pattern as other Postgres extensions, so I don't think any changes are needed (supporting 13, 15, and 16 would break existing Postgres installations - see homebrew-core for previous discussions / attempts at this).

> This is only the case if the user runs an unsupported brew configuration

Oh, my bad! I didn't realise those don't end up installed along pgvector itself. Do you have any links or keywords to find those previous discussions?

**Reply by ankane** [reply]
- https://github.com/Homebrew/homebrew-core/pull/181862
- https://github.com/Homebrew/homebrew-core/pull/167634
- https://github.com/Homebrew/homebrew-core/pull/159557
- https://github.com/Homebrew/homebrew-core/pull/159723

are a few of them.

**Reply by k0nserv** [reply]
Welp, I feel a bit silly for assuming the current lack of support for 13, 15, and 16 was just an oversight, I swear I tried to look around beforehand, apologies for the extra noise and thanks for providing context.
